src/kernel/sporadic.c

```diff
@@ -139,11 +145,15 @@ static inline void refill_add_tail(sched_context_t *sc, refill_t refill)
     assert(new_tail < sc->scRefillMax);
 }
 
+#ifdef ENABLE_SMP_SUPPORT
+static inline void maybe_add_empty_tail(sched_context_t *sc, word_t core)
+#else
 static inline void maybe_add_empty_tail(sched_context_t *sc)
+#endif
 {
     if (isRoundRobin(sc)) {
         /* add an empty refill - we track the used up time here */
-        refill_t empty_tail = { .rTime = NODE_STATE(ksCurTime)};
+        refill_t empty_tail = { .rTime = NODE_STATE_ON_CORE(ksCurTime, core)};
```

Comment:
Given this is only used for round-robin threads and the actual rTimes in a round-robin thread don't matter, it would be easier just to set it to be the same as the existing time (and not need to pass the core around).

```suggestion
        refill_t empty_tail = { .rTime = refill_head(sc)->rTime};
```

Reply:
Agree on it being an improvement to avoid passing the core around.

If set to the head like that does that maintain the invariant of ordered `rTime`? It's not immediately obvious to me that it does from just that context (where as cur time makes it clear that the invariant is maintained).

Specifically, the suggested change assumes that the list only has a single item at the head and/or the list already has all identical `rTime`.

Actually, I guess the post-check `assert(refill_size(sc) == MIN_REFILLS);` in effect asserts enforces that the pre-condition is that the list size is 1. I think that would be clearer with a pre-check of `assert(refill_size(sc) == 1);`

And, apologies for dragging this out, but is `MIN_REFILLS` an accurate name? It seems that `MIN_REFILLS` only applies to round-robin scheduling contexts. Would `MIN_RR_REFILLS` be considered as an alternative here?

Reply:
`MIN_REFILLS` applies to all lists of refills - we always need space to place the time left, and the consumed time. I'm not sure what the invariants look like now, but `rTime` is effectively ignored for RR threads. At one point I tried something clever to reduce branching and `rTime` did mean something for RR, but it made everything overcomplicated for little gain.

Reply:
OK, the only times I could find assertions were for round-robin. I'll need to read the code more deeply to actually understand what's going on.

So, putting `MIN_REFILLS` aside for now. I guess Curtis' suggested change makes sense. With a comment along the lines of:

*rTime is not used for round-robin scheduling context, however to maintain the invariant that the refill list is always ordered, set tail rTime to match header rTime*.

Reply:
Writing that out makes me think I've gone about this the wrong way. If `rTime` is actually no used for RR scheduling context. Then the invariant of *refill list should be ordered* should only apply to non-RR scheduling contexts.

So, any call of the nature `assert(refilled_ordered(sc))` should really be change to `assert(isRoundRobin(sc) || refill_ordered(sc))` and then `maybe_add_empty_tail` can simply add a zero-ed refill_t struct.

That seems slightly more elegant to me than initializing `rTime` to only satisfy an invariant that doesn't apply.

Reply:
There is another set of PRs that includes a change to `refill_ordered` that just returns `true` if the `isRoundRobin(sc)`.

https://github.com/seL4/seL4/pull/256/commits/7e2d92960ddd58cc2a5bddb8bc9399bbffd5343d#diff-819520826b8504ae909d5c66bbe6ede1R67-R69

Making just that change would also be an appropriate alternative. The abstract invariants now shouldn't actually depend on the `rTime` for round-robin SCs afaiaa.